benchmark failed

concurrent_map_bench_test.go
function BenchmarkMultiInsertDifferentSyncMap exist status 1

jeff@jeff-ryzen:~/go/go_concurrent-map/concurrent-map$ go version
go version go1.9 linux/amd64

$ go test -v -run="none" -bench="BenchmarkMultiInsert"
--- FAIL: BenchmarkMultiInsertDifferentSyncMap
goos: linux
goarch: amd64
BenchmarkMultiInsertDifferent_1_Shard-16      	  300000	     11303 ns/op
BenchmarkMultiInsertDifferent_16_Shard-16     	 2000000	       857 ns/op
BenchmarkMultiInsertDifferent_32_Shard-16     	 2000000	       904 ns/op
BenchmarkMultiInsertDifferent_256_Shard-16    	 1000000	      1556 ns/op
BenchmarkMultiInsertSame-16                   	  500000	      7450 ns/op
BenchmarkMultiInsertSameSyncMap-16            	  500000	      9161 ns/op
FAIL
exit status 1
FAIL	_/home/jeff/go/go_concurrent-map/concurrent-map	20.217s

--- a/README-zh.md
+++ b/README-zh.md
@@ -1,0 +1,64 @@
+# concurrent map [![Build Status](https://travis-ci.com/orcaman/concurrent-map.svg?branch=master)](https://travis-ci.com/orcaman/concurrent-map)
+
+正如 [这里](http://golang.org/doc/faq#atomic_maps) 和 [这里](http://blog.golang.org/go-maps-in-action)所描述的, Go语言原生的`map`类型并不支持并发读写。`concurrent-map`提供了一种高性能的解决方案:通过对内部`map`进行分片，降低锁粒度，从而达到最少的锁等待时间(锁冲突)
+
+在Go 1.9之前，go语言标准库中并没有实现并发`map`。在Go 1.9中，引入了`sync.Map`。新的`sync.Map`与此`concurrent-map`有几个关键区别。标准库中的`sync.Map`是专为`append-only`场景设计的。因此，如果您想将`Map`用于一个类似内存数据库，那么使用我们的版本可能会受益。你可以在golang repo上读到更多，[这里](https://github.com/golang/go/issues/21035) and [这里](https://stackoverflow.com/questions/11063473/map-with-concurrent-access)
+***译注:`sync.Map`在读多写少性能比较好，否则并发性能很差***
+
+## 用法
+
+导入包:
+
+```go
+import (
+	"github.com/orcaman/concurrent-map"
+)
+
+```
+
+```bash
+go get "github.com/orcaman/concurrent-map"
+```
+
+现在包被导入到了`cmap`命名空间下
+***译注:通常包的限定前缀(命名空间)是和目录名一致的，但是这个包有点典型😂，不一致！！！所以用的时候注意***
+
+## 示例
+
+```go
+
+	// 创建一个新的 map.
+	m := cmap.New()
+
+	// 设置变量m一个键为“foo”值为“bar”键值对
+	m.Set("foo", "bar")
+
+	// 从m中获取指定键值.
+	if tmp, ok := m.Get("foo"); ok {
+		bar := tmp.(string)
+	}
+
+	// 删除键为“foo”的项
+	m.Remove("foo")
+
+```
+
+更多使用示例请查看`concurrent_map_test.go`.
+
+运行测试:
+
+```bash
+go test "github.com/orcaman/concurrent-map"
+```
+
+## 贡献说明
+
+我们非常欢迎大家的贡献。如欲合并贡献，请遵循以下指引:
+- 新建一个issue,并且叙述为什么这么做(解决一个bug，增加一个功能，等等)
+- 根据核心团队对上述问题的反馈，提交一个PR，描述变更并链接到该问题。
+- 新代码必须具有测试覆盖率。
+- 如果代码是关于性能问题的，则必须在流程中包括基准测试(无论是在问题中还是在PR中)。
+- 一般来说，我们希望`concurrent-map`尽可能简单，且与原生的`map`有相似的操作。当你新建issue时请注意这一点。
+
+## 许可证
+MIT (see [LICENSE](https://github.com/orcaman/concurrent-map/blob/master/LICENSE) file)

--- a/README.md
+++ b/README.md
@@ -58,5 +58,8 @@ Contributions are highly welcome. In order for a contribution to be merged, plea
 - If the code is about performance issues, you must include benchmarks in the process (either in the issue or in the PR).
 - In general, we would like to keep `concurrent-map` as simple as possible and as similar to the native `map`. Please keep this in mind when opening issues.
 
+## language
+- [中文说明](./README-zh.md)
+
 ## license
 MIT (see [LICENSE](https://github.com/orcaman/concurrent-map/blob/master/LICENSE) file)

--- a/concurrent_map_bench_test.go
+++ b/concurrent_map_bench_test.go
@@ -89,10 +89,8 @@ func benchmarkMultiInsertDifferent(b *testing.B) {
 func BenchmarkMultiInsertDifferentSyncMap(b *testing.B) {
 	var m sync.Map
 	finished := make(chan struct{}, b.N)
-	err, set := GetSetSyncMap(&m, finished)
-	if err != nil {
-		b.FailNow()
-	}
+	_, set := GetSetSyncMap(&m, finished)
+
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
 		go set(strconv.Itoa(i), "value")
@@ -275,18 +273,20 @@ func GetSet(m ConcurrentMap, finished chan struct{}) (set func(key, value string
 		}
 }
 
-func GetSetSyncMap(m *sync.Map, finished chan struct{}) (set func(key, value string), get func(key, value string)) {
-	return func(key, value string) {
-			for i := 0; i < 10; i++ {
-				m.Load(key)
-			}
-			finished <- struct{}{}
-		}, func(key, value string) {
-			for i := 0; i < 10; i++ {
-				m.Store(key, value)
-			}
-			finished <- struct{}{}
+func GetSetSyncMap(m *sync.Map, finished chan struct{}) (get func(key, value string), set func(key, value string)) {
+	get = func(key, value string) {
+		for i := 0; i < 10; i++ {
+			m.Load(key)
 		}
+		finished <- struct{}{}
+	}
+	set = func(key, value string) {
+		for i := 0; i < 10; i++ {
+			m.Store(key, value)
+		}
+		finished <- struct{}{}
+	}
+	return
 }
 
 func runWithShards(bench func(b *testing.B), b *testing.B, shardsCount int) {

--- a/concurrent_map_test.go
+++ b/concurrent_map_test.go
@@ -71,7 +71,7 @@ func TestGet(t *testing.T) {
 	}
 
 	elephant, ok = tmp.(Animal) // Type assertion.
-	if &elephant == nil {
+	if !ok {
 		t.Error("expecting an element, not null.")
 	}
 

--- a/concurrent_map_test.go
+++ b/concurrent_map_test.go
@@ -65,14 +65,12 @@ func TestGet(t *testing.T) {
 	m.Set("elephant", elephant)
 
 	// Retrieve inserted element.
-
 	tmp, ok := m.Get("elephant")
-	elephant = tmp.(Animal) // Type assertion.
-
 	if ok == false {
 		t.Error("ok should be true for item stored within the map.")
 	}
 
+	elephant, ok = tmp.(Animal) // Type assertion.
 	if &elephant == nil {
 		t.Error("expecting an element, not null.")
 	}

